pkg/controllermanager/controller/shoot/shoot_hibernation_control.go

```diff
@@ -145,110 +57,292 @@ func (c *Controller) shootHibernationAdd(obj interface{}) {
+		return err
 	}
 
-	r.hibernationScheduleRegistry.Delete(key)
-	logger.Debugf("Deleted cron")
+	patch = client.MergeFrom(shoot.DeepCopy())
+	shoot.Status.LastHibernationTriggerTime = &metav1.Time{Time: now}
+	return r.gardenClient.Status().Patch(ctx, shoot, patch)
 }
 
-func (r *shootHibernationReconciler) createOrUpdateShootCron(ctx context.Context, logger logrus.FieldLogger, key string, shoot *gardencorev1beta1.Shoot) error {
-	r.deleteShootCron(logger, key)
-	if !shootHasHibernationSchedules(shoot) {
-		return nil
+// parseHibernationSchedules parses the given HibernationSchedules and returns an array of ParsedHibernationSchedules
+// If the Location of a HibernationSchedule is `nil`, it is defaulted to UTC.
+func parseHibernationSchedules(schedules []gardencorev1beta1.HibernationSchedule) ([]parsedHibernationSchedule, error) {
+	var parsedHibernationSchedules []parsedHibernationSchedule
+
+	for _, schedule := range schedules {
+		var locationID string
+		if schedule.Location != nil {
+			locationID = *schedule.Location
+		} else {
+			locationID = time.UTC.String()
+		}
+
+		location, err := time.LoadLocation(locationID)
+		if err != nil {
+			return nil, err
+		}
+		if schedule.Start != nil {
+			parsed, err := cron.ParseStandard(*schedule.Start)
+			if err != nil {
+				return nil, err
+			}
+			parsedHibernationSchedules = append(parsedHibernationSchedules,
+				parsedHibernationSchedule{location: location, schedule: parsed, operation: hibernate},
+			)
+		}
+		if schedule.End != nil {
+			parsed, err := cron.ParseStandard(*schedule.End)
+			if err != nil {
+				return nil, err
+			}
+			parsedHibernationSchedules = append(parsedHibernationSchedules,
+				parsedHibernationSchedule{location: location, schedule: parsed, operation: wakeUp},
+			)
+		}
 	}
 
-	schedule, err := ComputeHibernationSchedule(ctx, r.gardenClient, logger, r.recorder, shoot)
-	if err != nil {
-		return err
+	return parsedHibernationSchedules, nil
+}
+
+// nextHibernationTimeDuration returns the time duration after which to requeue the shoot based on the hibernation schedules and current time.
+// It adds a 100ms padding to the next requeue to account for Network Time Protocol(NTP) time skews.
+// If the time drifts are adjusted which in most realistic cases would be around 100ms, scheduled hibernation
+// will still be executed without missing the schedule.
+func nextHibernationTimeDuration(schedules []parsedHibernationSchedule, now time.Time) time.Duration {
+	var timestamps []time.Time
+	for _, schedule := range schedules {
+		ts := schedule.next(now)
+		timestamps = append(timestamps, ts)
 	}
 
-	schedule.Start()
-	r.hibernationScheduleRegistry.Store(key, schedule)
-	logger.Debugf("Successfully started hibernation schedule")
+	sort.Slice(timestamps, func(i, j int) bool {
+		return timestamps[i].Before(timestamps[j])
+	})
+
+	duration := timestamps[0].Add(nextScheduleDelta).Sub(now)
+	return duration
```

Comment:
```suggestion
	return timestamps[0].Add(nextScheduleDelta).Sub(now)
```

Reply:
Can `timeStamp`s be `nil` or empty?

Reply:
The code will not get to this point, since there is a check whether the shoot contains any hibernation schedules before this function is called. There's a test case for reconciling a shoot without hibernation schedules.

Reply:
OK, thanks